AndroidTest/app/src/main/java/cloudant/com/androidtest/keyprovider/AndroidKeyProviderTests.java

```diff
@@ -0,0 +1,173 @@
+                    .getSimpleName() + ": " + t.getLocalizedMessage());
+        }
+    }
+
+    public void testCreateProviderWithNullPassword() {
+        try {
+            new AndroidKeyProvider(getContext(), null, ProviderTestUtil.getUniqueIdentifier());
+            fail("AndroidKeyProvider constructor should fail if password is null");
+        } catch (IllegalArgumentException e) {
+            assertNotNull(e);
+        } catch (Throwable t) {
+            fail("Failed to throw IllegalArgumentException.  Found: " + t.getClass()
+                    .getSimpleName() + ": " + t.getLocalizedMessage());
+        }
+    }
+
+    public void testCreateProviderWithEmptyPassword() {
+        try {
+            new AndroidKeyProvider(getContext(), "", ProviderTestUtil.getUniqueIdentifier());
+            fail("AndroidKeyProvider constructor should fail if password is empty string");
+        } catch (IllegalArgumentException e) {
+            assertNotNull(e);
+        } catch (Throwable t) {
+            fail("Failed to throw IllegalArgumentException.  Found: " + t.getClass()
+                    .getSimpleName() + ": " + t.getLocalizedMessage());
+        }
+    }
+
+    public void testCreateProviderWithNullIdentifier() {
+        try {
+            new AndroidKeyProvider(getContext(), "passw0rd", null);
+            fail("AndroidKeyProvider constructor should fail if identifier is null");
+        } catch (IllegalArgumentException e) {
+            assertNotNull(e);
+        } catch (Throwable t) {
+            fail("Failed to throw IllegalArgumentException.  Found: " + t.getClass()
+                    .getSimpleName() + ": " + t.getLocalizedMessage());
+        }
+    }
+
+    public void testCreateProviderWithEmptyIdentifier() {
+        try {
+            new AndroidKeyProvider(getContext(), "password", "");
+            fail("AndroidKeyProvider constructor should fail if identifier is empty string");
+        } catch (IllegalArgumentException e) {
+            assertNotNull(e);
+        } catch (Throwable t) {
+            fail("Failed to throw IllegalArgumentException.  Found: " + t.getClass()
+                    .getSimpleName() + ": " + t.getLocalizedMessage());
+        }
+    }
+
+    public void testLoadWithWrongPassword() {
+        String identifier = ProviderTestUtil.getUniqueIdentifier();
+        AndroidKeyProvider provider1 = new AndroidKeyProvider(getContext(), "provider1password",
+                identifier);
+        AndroidKeyProvider provider2 = new AndroidKeyProvider(getContext(), "wrongpassword",
+                identifier);
+
+        EncryptionKey provider1Key = provider1.getEncryptionKey();
+        assertNotNull("provider1 EncryptionKey was not created", provider1Key);
+
+        try {
+            provider2.getEncryptionKey();
+            fail("AndroidKeyProvider getEncryptionKey should fail if password is not correct");
+        } catch (DPKException e) {
```

Comment:
There are other reasons that these operations could fail.  These include InvalidKeySpecException, IllegalBlockSizeException, InvalidKeyException, BadPaddingException, BadPaddingException, NoSuchPaddingException.  We could create a BadPasswordException class to capture some of these.  Others are not caused by a bad password.  I used the DPKException to surface all of these and embedded the cause exception inside of it.